modules/videoio/src/cap_gstreamer.cpp

```diff
@@ -609,6 +835,20 @@ bool GStreamerCapture::retrieveFrame(int, OutputArray dst)
     CV_Error_(Error::StsNotImplemented, ("Unsupported GStreamer layer type: %s", name.c_str()));
 }
 
+bool GStreamerCapture::retrieveFrame(int index, OutputArray dst)
+{
+    if (index < 0)
+        return false;
+    if (!sample)
+        return false;
+
+    if (videoStream != -1)
+        return retrieveVideoFrame(index, dst);
+    else if (audioStream != -1)
+        return retrieveAudioFrame(index, dst);
```

Comment:
This logic is not correct. Checks must be "index"-based.

What is the problem to write correct logic once instead of rewriting them again with adding video support?